llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -426,6 +429,32 @@ static unsigned getMinSchedulableCycle(SchedBoundary &Zone) {
   return MinSchedulableCycle;
 }
 
+bool AIEPostRASchedStrategy::doesNotProgressInZone(const SchedBoundary &Zone,
+                                                   const SUnit &SU) const {
+  // If SU is a fixed instruction in the other zone, it isn't available.
+  if (isFixedSU(SU, /*IsTop*/ !Zone.isTop()))
+    return true;
+
+  // We cannot proceed with delay slot instructions in the top zone.
+  return Zone.isTop() && SU.getInstr()->hasDelaySlot();
+}
+
+// This function returns true when it is impossible to continue with top-down
+// without entering an infinite loop because the only remaining instructions
+// cannot be scheduled in the top zone.
+bool AIEPostRASchedStrategy::mustSwitchToBottomUp() {
```

Comment:
How about never adding the delay slot instruction to the Top zone instead? We could override `AIEPostRASchedStrategy::releaseTopNode()` so that the instruction does not get added to the queues of `Top` if it has delay slots.